Trace (#19)
yiisoft/yii-sentry: README.md

];
```

If you want to trace Guzzle requests and add Sentry headers to external queries, add the following:

```php 
GuzzleHttp\Client::class => static function (ContainerInterface $container) {
    $stack = new HandlerStack();
    $stack->setHandler(new CurlHandler());
    $factory = $container->get(GuzzleMiddlewareFactory::class);
    $middleware = static function (callable $handler) use ($factory): callable {
        return $factory->factory($handler);
    };

    $stack->push($middleware);

    return new GuzzleHttp\Client([
        'handler' => $stack,
    ]);
},
```

**Configure:**

Add the following code block to your `params.php` and define DSN. Also you can set "environment" and "release". Good example is to use TAG from gitlab.ci for it.
```php 
'yiisoft/yii-sentry' => [
    'options' => [
        'dsn' => '',
        'environment' => 'local', //SENTRY_ENVIRONMENT, //YII_ENV,
        'release' => 'dev',  //SENTRY_RELEASE, //TAG

>         // @see: https://docs.sentry.io/platforms/php/configuration/options/#send-default-pii

Reviewer: ```suggestion
        // https://docs.sentry.io/platforms/php/configuration/options/#send-default-pii
```

Reviewer: This tag has no effect here and it's intended more for linking methods / properties, not for external links.

        'send_default_pii' => true,
        'traces_sample_rate' => 1.0,
    ],
    'handleConsoleErrors' => true,
    'log_level' => 'warning',
    'tracing' => [
        // Indicates if the tracing integrations supplied by Sentry should be loaded
        'default_integrations'   => true,
    ],
]
```

Add `APP_START_TIME` constant into `index.php` and `yii.php`:

```php
define('APP_START_TIME', microtime(true));
```

Add log targets for breadcrumbs and tracing to `app/config/common/logger.php` or another config file with logger settings:

> ```php 

Reviewer: ```suggestion
```php
```

];
```

> if you want to trace guzzle requests and add sentry headers to external queries, add this

Reviewer 2: ```suggestion
If you want to trace Guzzle requests and add Sentry headers to external queries, add the following:
```

```php 

>     GuzzleHttp\Client::class => static function (ContainerInterface $container) {
>         $stack = new HandlerStack();
>         $stack->setHandler(new CurlHandler());
>         $factory = $container->get(GuzzleMiddlewareFactory::class);
>         $middleware = static function (callable $handler) use ($factory): callable {
>             return $factory->factory($handler);
>         };
> 
>         $stack->push($middleware);
> 
>         return new GuzzleHttp\Client([
>             'handler' => $stack,
>         ]);
>     },

Reviewer 2: ```suggestion
GuzzleHttp\Client::class => static function (ContainerInterface $container) {
    $stack = new HandlerStack();
    $stack->setHandler(new CurlHandler());
    $factory = $container->get(GuzzleMiddlewareFactory::class);
    $middleware = static function (callable $handler) use ($factory): callable {
        return $factory->factory($handler);
    };

    $stack->push($middleware);

    return new GuzzleHttp\Client([
        'handler' => $stack,
    ]);
},
```

```

**Configure:**

> add code block below to your params.php
> and type your DSN
> also you can define your environment and release, for example TAG from gitlab.ci

Reviewer 2: ```suggestion
Add the following code block to your `params.php` and define DSN. Also you can set "environment" and "release". Good example is to use TAG from gitlab.ci for it.
```

```php 

>     'yiisoft/yii-sentry' =>
>         [
>             'options' => [
>                 'dsn' => '',
>                 'environment' => 'local', //SENTRY_ENVIRONMENT, //YII_ENV,
>                 'release' => 'dev',  //SENTRY_RELEASE, //TAG
>                 // @see: https://docs.sentry.io/platforms/php/configuration/options/#send-default-pii
>                 'send_default_pii' => true,
>                 'traces_sample_rate' => 1.0,
>             ],
>             'handleConsoleErrors' => true,
>             'log_level' => 'warning',
>             'tracing'          => [
>                 // Indicates if the tracing integrations supplied by Sentry should be loaded
>                 'default_integrations'   => true,
>             ],
>         ]

Reviewer 2: ```suggestion
'yiisoft/yii-sentry' => [
    'options' => [
        'dsn' => '',
        'environment' => 'local', //SENTRY_ENVIRONMENT, //YII_ENV,
        'release' => 'dev',  //SENTRY_RELEASE, //TAG
        // @see: https://docs.sentry.io/platforms/php/configuration/options/#send-default-pii
        'send_default_pii' => true,
        'traces_sample_rate' => 1.0,
    ],
    'handleConsoleErrors' => true,
    'log_level' => 'warning',
    'tracing' => [
        // Indicates if the tracing integrations supplied by Sentry should be loaded
        'default_integrations'   => true,
    ],
]
```

```

> add APP_START_TIME const into index.php and yii.php

Reviewer 2: ```suggestion
Add `APP_START_TIME` constant into `index.php` and `yii.php`:

```

```php
define('APP_START_TIME', microtime(true));
```

> add log targets for breadcrumbs and tracing in app/config/common/logger.php
> or another config file with logger settings

Reviewer 2: ```suggestion
Add log targets for breadcrumbs and tracing to `app/config/common/logger.php` or another config file with logger settings:
```

```php 
return [
    LoggerInterface::class => static function (
        /** your_another_log_target $your_log_target */
        \Yiisoft\Yii\Sentry\SentryBreadcrumbLogTarget $sentryLogTarget,
        Yiisoft\Yii\Sentry\Tracing\SentryTraceLogTarget $sentryTraceLogTarget
    ) {
        return new Logger([
        /** $your_log_target */
            $sentryLogTarget,
            $sentryTraceLogTarget
        ]);
    }
];
```

> **if you want to see your logs in sentry timeline**, you need to use keys (float)'**time**' and (float)'**elapsed**' in log context array

Reviewer 2: ```suggestion
> Note: **If you want to see your logs in sentry timeline**, you need to use keys (float)'**time**' and (float)'**elapsed**' in log context array.
```

> _____

Reviewer 2: ```suggestion
```

> add DB log decorator for tracing db queries in app/config/params.php

Reviewer 2: ```suggestion
Add DB log decorator for tracing db queries in `app/config/params.php`:

```

```php

> 'yiisoft/yii-cycle' => [
>         // DBAL config
>         'dbal' => [
>             // SQL query logger. Definition of Psr\Log\LoggerInterface
>             // For example, \Yiisoft\Yii\Cycle\Logger\StdoutQueryLogger::class
>             'query-logger' => \Yiisoft\Yii\Sentry\DbLoggerDecorator::class,
>             /**
>             * ...
>             * your another db settings 
>             **/
>     ]
> ]

Reviewer 2: ```suggestion
'yiisoft/yii-cycle' => [
    // DBAL config
    'dbal' => [
        // SQL query logger. Definition of Psr\Log\LoggerInterface
        // For example, \Yiisoft\Yii\Cycle\Logger\StdoutQueryLogger::class
        'query-logger' => \Yiisoft\Yii\Sentry\DbLoggerDecorator::class,
        /**
         * ...
         * your another db settings 
         **/
    ]
]
```

```

> add into app/config/params.php into middleware section  SetRequestIpMiddleware

Reviewer 2: ```suggestion
Add `SetRequestIpMiddleware` to `app/config/params.php`, "middleware" section:
```

```php
    'middlewares' => [
        ErrorCatcher::class,
        \Yiisoft\Yii\Sentry\Http\SetRequestIpMiddleware::class, //add this
        Router::class,
    ],
```

> add into app/config/common/router.php tracing middleware

Reviewer 2: ```suggestion
Add `SentryTraceMiddleware` to `app/config/common/router.php`:
```

```php
  RouteCollectionInterface::class => static function (RouteCollectorInterface $collector) use ($config) {
        $collector
            ->middleware(FormatDataResponse::class)
            ->middleware(JsonParseMiddleware::class)
            ->middleware(ExceptionMiddleware::class)
            ->middleware(\Yiisoft\Yii\Sentry\Tracing\SentryTraceMiddleware::class) // add this
            ->addGroup(
                Group::create('')
                    ->routes(...$config->get('routes'))
            );

        return new RouteCollection($collector);
    },
 ```

> ________

Reviewer 2: ```suggestion
```

> 
> 
> 
> if your transaction too heavy you can slice it to several transactions with clearing log buffer.
> 
> use SentryConsoleTransactionAdapter or SentryWebTransactionAdapter
> 
> for example:

Reviewer 2: ```suggestion

If your transaction is too heavy, you can slice it to several transactions with clearing log buffer. Use `SentryConsoleTransactionAdapter` or `SentryWebTransactionAdapter`. For example:
```

```php

>         /** some code with default transaction */
>         /** commit default transaction and send data to sentry server */
>         $sentryTraceString = $this->sentryTransactionAdapter->commit();
>         while ($currentDate <= $endDate) {
>             $this->sentryTransactionAdapter->begin($sentryTraceString)
>                 ->setName('my_heavy_operation/iteration')
>                 ->setData(['date' => $currentDate->format('Y-m-d')]);
> 
>             $this->process($currentDate, $sentryTraceString);
>             $this->sentryTransactionAdapter->commit();
>         }
>         $this->sentryTransactionAdapter->begin($sentryTraceString)
>             ->setName('my_heavy_operation done, terminating application');
>     /** transaction will commit when application is terminated */

Reviewer 2: ```suggestion
/** some code with default transaction */
/** commit default transaction and send data to sentry server */
$sentryTraceString = $this->sentryTransactionAdapter->commit();
while ($currentDate <= $endDate) {
    $this->sentryTransactionAdapter->begin($sentryTraceString)
        ->setName('my_heavy_operation/iteration')
        ->setData(['date' => $currentDate->format('Y-m-d')]);

    $this->process($currentDate, $sentryTraceString);
    $this->sentryTransactionAdapter->commit();
}

$this->sentryTransactionAdapter->begin($sentryTraceString)
    ->setName('my_heavy_operation done, terminating application');
/** transaction will commit when application is terminated */
```

```

> for this example all new transactions will linked to transaction with $sentryTraceString

Reviewer 2: ```suggestion
In this example all new transactions will linked to transaction with `$sentryTraceString`.
```
